### PR TITLE
[LLVMCPU] Add support for using multiple lowering_config.

### LINKED ISSUE #13795: Takes forever to compile MobileNetV2 Int8 on x86 using data tiling
I hit the issue when trying more things with data tiling. File an issue for tracking it... To repro, import the model from 

https://github.com/openxla/iree/blob/fa5c6b9edcfbc0a37fb144d916357245a505ec19/build_tools/python/e2e_test_framework/models/tflite_models.py#L133-L142

Run `iree-compile --output-format=vm-bytecode --iree-hal-target-backends=llvm-cpu --iree-input-type=tosa int8.mlir -o /tmp/a.vmfb --iree-llvmcpu-target-triple=x86_64-pc-linux-gnu --iree-llvmcpu-target-cpu-features=host`

### PR DESCRIPTION
**Pull request by hanhanW**
Some dimensions might not be tiled because

  1. They are not part of root op dimensions.
  2. The lowering_config disappears after SplitReduction

An example is reduction + broadcast. This commit adds customized
lowering_config to all the compute ops. The tile sizes are identical
except inner level of tiling. I.e., the distribution and TileAndFuse
tile sizes are identical. The rest tile sizes depend on the op itself.

Fixes https://github.com/openxla/iree/issues/13795

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: This line does seem to match the comment above. Is this supposed to be `parallelTileSizes`?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:2095]: Maybe some comments about what is happening here?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:2121]: It might be better to move all of this out of the `Case` statement itself into a function, and document exactly what is happening here. This is fairly non-trivial?

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Thanks! I think mostly for looking for more comments in the code to describe what exactly is happening.....

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: nit: could we replace the `XYZTileSizes.size()` with `numTilingDims` or similar? That would make it easy to read and more explicit about any assumptions needed to understand this code.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:1179]: doc?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: Wondering if we really have to enforce a 4th level for all the ops given that the last one is not applicable to most of them. We can have a 4th level only for the ops that need it. `TilingConfig` should be able to handle configurations with 3, 4 and 5 levels and do the proper verification to make sure that each configuration is valid. WDYT?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: Why are we using empty tuples instead of the ones with zeros here?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (dcaballe) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:2095]: +1

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (dcaballe) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:2121]: yep, that would be great. There is also some code that is common and could be refactored.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:464]: Not sure I understand the copy in 451. We should also take advantage of the fact that fusable levels are contiguous and avoid lookups (what the previous code does).

File: compiler/src/iree/compiler/Codegen/LLVMCPU/TileSizeSelection.cpp
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/TileSizeSelection.cpp:28]: I would like to better understand what this inner parallel level represents. Do you have an example?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Utils.h
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/Utils.h]: typo

**Review by dcaballe — COMMENTED** [review]
Looks good but I think I need a bit more time to digest this. Some comments for now!

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: They are not applicable to most of them because they are reduction ops and they don't have parallel dims other than common parallel ones. The current implementation is:

1. We find a root op from compute ops.
2. We set lowering_config for the root op.
3. We dup distribution and common parallel TileAndFuse list to other compute ops.
4. We set tile sizes for the rest of dims for all the computation ops (except root op because that's already set in (2)).

Maybe we can model it in a new way (which needs much more changes):

1. Given that we have a sequence of compute op, we can extract common parallel dims and set distribution tile sizes and first TileAndFuse tile sizes.
2. We iterate all the compute ops and set the tile sizes for the rest of dims.

As we've discussed long time ago, we decide to go with the former implementation for now. So I ended up with the commit.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: Setting tiling size to zeros is basically disabling tiling? The op does not have reduction dims, so the third list can't be `parallelTileSizes`.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: I was just following what had been used in this method. They have the same behavior. I agree that we should be consistent, so I update it to zeros.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:464]: They could be non-contiguous, see the example in `setLoweringConfigForComputeOps` comment. The TileAndFuse can happen in the fourth tile list for elementwise + pack ops. E.g., IR after tiling reduction dims on reduction ops:

```
scf.for ...
  reduction
elementwise
pack
```

We are able to tile and fuse elementwise + pack ops using the fourth tile list:

```
scf.for ...
  reduction
scf.for ..
  elementwise
  pack
```

File: compiler/src/iree/compiler/Codegen/LLVMCPU/TileSizeSelection.cpp
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/TileSizeSelection.cpp:28]: See the example in setLoweringConfigForComputeOps comment

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:2095]: done, please take a look

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:2121]: I will do it after I address other concerns

**Review by MaheshRavishankar — APPROVED** [review]
Ok, looks good to me. Please wait for Diego's review.

**Comment by dcaballe** [comment]
There are some large regressions on x86. Is that expected?

**Comment by hanhanW** [comment]
> There are some large regressions on x86. Is that expected?

No, it's not expected. The improvements were from https://github.com/openxla/iree/pull/13340, and now it regresses in the PR. I'll take a look at it.

Besides the regression, do you have concerns about how we structure TilingConfig?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: typo

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:2075]: What would happen if we had more dimensions here? E.g., ["parallel", "parallel", "parallel", "parallel"]

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (dcaballe) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:2095]: Thanks! That's very helpful

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: Could you please add a note/example about why this is needed? Otherwise, it will be difficult to get rid of it/investigate this further.

**Review by dcaballe — APPROVED** [review]
We can move forward with this, thanks, and sorry for the long turn around. I'm not sure about the impact on the ongoing multi-level tiling code... Will see. Please, take a look at the comments before landing. THanks!

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:2075]: Say that the lowering_config of rootOp is `[[X, 0], [Y, 0], [0, Z], [0, 0]]`, it will have [[X, 0, 0, 0], [Y, 0, 0, 0], [0, 0, 0, 0], [0, A, B, C]]`.

They have shared tile sizes for distribution and TileAndFuse, so `X` and `Y` are used in the first two list. There are no reduction dims, so the third list is empty. The `[0, A, B, C]` is derived from the characteristic of elementwise op (e.g., `getMinTilingSizesForEachDim`).

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:2121]: The common part is ~10 lines here, and the refactoring would have more than 100 lines changes. Because there are three generic op configuration methods. I will do it as a follow-up, so we can have less changes this PR. (it's already quite big.)

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: done

**Comment by hanhanW** [comment]
> We can move forward with this, thanks, and sorry for the long turn around. I'm not sure about the impact on the ongoing multi-level tiling code... Will see. Please, take a look at the comments before landing. THanks!

Thanks for reviewing it and the flexibility! I can definitely help on conflicts and address the impact on multi-level tiling code when needed!

**Comment by bjacob** [comment]
@hanhanW @mariecwhite @MaheshRavishankar, this also fixed #14414  !

**Pull request merged**
